### ISSUE TITLE
Getting the card brand name

### ISSUE THREAD

**Issue by yazinsai** [Fivell/credit_card_validations#18]
```ruby
require 'credit_card_validations/string'

'4242 4242 4242 4242'.brand # => :visa
puts '4242 4242 4242 4242'.brand # => "visa"
```

But how do I get the readable credit card brand you've listed in the readme i.e. `"Visa"` for `:visa`, or `"American Express"` for `:amex`?

**Comment by Fivell** [comment]
@yazinsai , there is no "readable" strings inside gem like in README. 
but you can use locales for this if you need

**Comment by yazinsai** [comment]
Thanks for the suggestions @Fivell, I'll be using them for my project. 

For the benefit of other users, I think would be great to have it in the gem itself (in case other cards are later added). 

Would you like me to send in a PR?

**Comment by Fivell** [comment]
@yazinsai , I agree that  some sort of "full brand name"  can be useful in some cases. 
do you want to implement something like ?
```ruby
'4242 4242 4242 4242'.brand_name # "Visa", "American Express", etc
```
BTW, PR are always welcome !

**Comment by yazinsai** [comment]
@Fivell exactly!

**Comment by yazinsai** [comment]
@Fivell a few questions I'll need your help with:

1. Where should the brand names be stored? I initially thought of including them in `card_rules.rb` since they're all listed there, but that seems like a conflict of concerns. Would like your input on that.

2. Would be good with a backwards-incompatible change to `add_rule`? 
Before: `add_rule(brand, length, prefixes, skip_luhn = false)`
After: `add_rule(brand, length, prefixes, brand_name, skip_luhn = false)`

**Comment by Fivell** [comment]
@yazinsai,  ```card_rules.rb``` is just storage for initial data (in future it should be plain yaml file) I think there is no other place to store new info, so it is good enough for now. 

About  ```add_rule method```, it starting to  smell because of too many arguments count.


Maybe it is better to use options  hash to incapsulate arguments that don't contain detecting logic.
Sort of
```add_rule(brand, length, prefixes, options)```

 options can contain not mandatory attributes: 
    ```skip_luhn``` - # treated as ```false``` if was not  passed
   ```brand_name``` # just for overriding default behaviour which is capitalizing first letter. 
  
So it is not needed to pass Visa for visa and Maestro for maestro , etc .  But American Express for amex and JCB for jcb, etc.
 
Also I don't think it can be backwards-compatible,  also method become more generic so maybe it will be better to create ```add_brand``` method and deprecate  ```add_rule```

To summarize all the changes,  version of gem will be 2.0.0 so no  backwards-compatiblity is guaranteed

**Comment by Fivell** [comment]
one more thing that I want to change is passing length , prefixes pair as key value hash. Because card prefix rule can be different in scope of length rule. See example
https://github.com/Fivell/credit_card_validations/blob/master/lib/credit_card_validations/card_rules.rb#L34

So maybe add_brand can have next  arguments
```ruby 
add_brand(key, rules , options = {})
```

 where rules is hash like 
```ruby
{ 
 [15,16] => ['3528', '3529', '353', '354', '355', '356', '357', '358'],
 [15] => ['1800', '2131']
}
```

or can be array like in initial rules CardRules module.

Like
```ruby
[
      {length: [15,16], prefixes: ['3528', '3529', '353', '354', '355', '356', '357', '358']},
      {length: [15], prefixes: ['1800', '2131']}
]
```
In this case we can have 2 methods add_rule and add_brand

```add_brand``` can create shortcut https://github.com/Fivell/credit_card_validations/blob/master/lib/credit_card_validations/detector.rb#L57
and then call add_rule for each length->prefixes pair. 

Looks like this simple issue turned to be trigger for good refactoring.....
What do you think about all of this @yazinsai   ?

**Comment by Fivell** [comment]
@yazinsai , I will create new branch 2.0.0 and I will start refactoring there

**Comment by yazinsai** [comment]
@Fivell Glad to help. I'll go ahead and make the changes we agreed to earlier, and then you can continue the refactoring to include the length / rules changes.

**Comment by Fivell** [comment]
@yazinsai , I will merge refactoring first, because there to much refactored code . After this we can work on other features simultaneously

**Comment by Fivell** [comment]
https://github.com/Fivell/credit_card_validations/issues/22 is blocker for this task

**Comment by Fivell** [comment]
@yazinsai , do you need any help?

**Comment by Fivell** [comment]
@yazinsai , do you still want to implement this ?

**Comment by yazinsai** [comment]
@Fivell Sorry, missed your earlier hints at completion. Working on this now..
